textproc/cdiff/src/cdiff.1
.\"-
.\" Copyright (c) 2000 Eivind Eklund
.\" All rights reserved.
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer
.\"    in this position and unchanged.
.\" 2. Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in the
.\"    documentation and/or other materials provided with the distribution.
.\" 3. The name of the author may not be used to endorse or promote products
.\"    derived from this software without specific prior written permission.
.\"
.\" THIS SOFTWARE IS PROVIDED BY THE AUTHOR ``AS IS'' AND ANY EXPRESS OR
.\" IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES
.\" OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED.
.\" IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY DIRECT, INDIRECT,
.\" INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT
.\" NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
.\" DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
.\" THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
.\" (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF
.\" THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
.\"
.\"	$FreeBSD$
.\"
.Dd October 24, 2000
.Dt CDIFF 1
.Sh NAME
.Nm cdiff
.Nd show diffs with colors
.Sh SYNOPSIS
.Nm cdiff
.Op Fl m
.Op Ar diff ...
.Sh DESCRIPTION
.Nm
reads a unidiff or context diff from stdin or a set of (optionally compressed)
files and URLs, adds color codes, expands tabs to 8 spaces, and passes the
results to
.Xr less 1 .
If used on a color terminal, it makes diffs much easier to read.
There is also handling of "cvs annotate" in it.
.Pp
.Bl -tag -width Ds
.It Fl m
Use bold, reverse and underline codes (for monochrome terminals).
.El
.Bl -tag -width CDIFFCOLORS
.It Ev CDIFFCOLORS
The color codes used to add. Default is 1:34:31:35, for lines begin with "@", "-", "+" and "!".
.El
.Sh AUTHORS
.Nm
was written by
.An Eivind Eklund Aq eivind@FreeBSD.org .
based on a sed line somebody passed around.  The code is (much) smaller than
this man page.
.Sh SEE ALSO
.Xr diff 1 ,
.Xr less 1 ,
.Xr cvs 1 .
.Sh BUGS
.Nm
assumes standard ANSI color escape sequences (as used by xterm and most other
commonly used terminals), rather than using
.Xr curses 3 .
.Pp
.Nm
should really be a large and complex C program based on
.Xr less 1
and a diff library.  This would allow a more informative display, including
collapsing pure deletions/insertions inside a single line to one line with
colors for deleted/added material, avoiding the line-wrap problems in less
due to less not knowing the effects or lengths of the color escape codes,
and correct handling of shifted tabs (instead of just expanding them to
spaces.)


CHANGES
Updating Information for FreeBSD ports developers

This file is maintained by Erwin Lansing <erwin@FreeBSD.org> and
copyrighted by the FreeBSD Foundation.

This file contains major changes to ports and the ports infrastructure.
Intended audience are ports committers, maintainers and other
developers. User oriented changes should be submitted for inclusion
in the release notes.

All ports committers are allowed to commit to this file.

20061014:
AUTHOR: gnome@FreeBSD.org
The following changes have been introduced to the GNOME porting process
during the GNOME 2.16 update:

* GNOME is now installed into LOCALBASE.  Any port which makes use of the
  gnomeprefix component will have its PREFIX forced to LOCALBASE

20060930:
AUTHOR: portmgr@FreeBSD.org
The following functional changes were made to bsd.port.mk:

  * The bsd.port.mk pre and post includes have been split into 3 pieces
    instead of 2, to allow OPTIONS to be able to influence dependencies.
    This is still experimental and not yet enabled by any port.  This adds
    the file 'bsd.port.options.mk'.

  * bsd.gcc.mk has been updated to understand gfortran.

  * emulators/linux_base is now removed.  linux_base-fc4 has been the default
    for some time.

  * The USE_FIREBIRD macro was added to bsd.database.mk.

20060914:
AUTHOR: portmgr@FreeBSD.org
The following functional changes were made to bsd.port.mk:

  * PKGVERSION was introduced as an intermediate variable refactored out of
    PKGNAME.

  * bsd.lua.mk was added to support lua-related ports.

  * DESTDIR was removed from bsd.scons.mk because it was problematic.

  * Some quotes were added to the 'missing' target to improve handling of
    duplicate origins.

20060903:
AUTHOR:	sat@FreeBSD.org
The following functional changes were made to bsd.sites.mk:

  * MASTER_SITES variable can now look like MACRO1/subdir1 MACRO2/subdir2, and
    be expanded in a way that you expect it to, that is apply different subdirs
    to different sites. You can safely replace ${M_S_FOO:S/%SUBDIR%/bar/} with
    FOO/bar.

  * If M_S_SUBDIR is unset, a macro in M_S won't set it anymore, but rather
    substitute %SUBDIR% all by itself. Now it's possible to use macros with
    different default subdirs and they will expand in a proper way.

20060805:
AUTHOR:	sat@FreeBSD.org
The following functional changes were made to bsd.sites.mk:

  * MASTER_SITES variable can now contain special macros which are expanded into
    traditional URL[:group] syntax and cause some convenient effects, e.g. set a
    default M_S_SUBDIR.

  * A macro is any space-delimited word in MASTER_SITES without a forward slash.

  * If MASTER_SITE_<macro> is defined, then the macro is expanded to its value.

  * Abbreviated macros are provided for extremely popular locations, e.g. CPAN
    and SF. To avoid further obfuscation their unabbreviated equivalents should
    be avoided.

  * Popular macros can trigger M_S_SUBDIR to default to a predefined value, e.g.
    ${PORTNAME:L} for SF and ${PORTNAME:C/-.*//} for CPAN.

  * Abbreviations and M_S_SUBDIR defaults are defined by MASTER_SITES_ABBREVS
    and MASTER_SITES_SUBDIRS, both set in bsd.sites.mk.

  * If a macro belongs to a group (e.g. SF:source1), all the sites it expands
    into also belong to the same group.

  * Macros and traditional URL's can be mixed safely, their order stays intact.

20060804:
AUTHOR:	gabor@FreeBSD.org
The following functional changes were made to bsd.*.mk and to 
Tools/scripts/security-check.awk:

  * Add DESTDIR support to let one install ports into a jail from outside.
    A package is installed under ${DESTDIR}${PREFIX} now and registered under
    ${DESTDIR}/var/db/pkg.

  * TARGETDIR variable has been added for referencing ${DESTDIR}${PREFIX}.

  * LOCALBASE, LINUXBASE and X11BASE variables point to the final destination
    now, LOCALBASE_REL, LINUXBASE_REL and X11BASE_REL can be used for the
    old behavior.  The behavior has not been changed in PLIST_SUB.
    Note, that the three variables are still overrideable with a hack, so this
    change doesn't affect POLA, but necessary for DESTDIR support.

  * OSVERSION is now determined from the userland, not from the kernel.
    Cross-compiling between releases is still not supported, but this
    behavior is more correct for jails with differing userland from
    the kernel version.

  * Add some sanity check for PREFIX and DESTDIR.  This can be skipped by
    defining IGNORE_PATH_CHECKS, but very discouraged.

  * Deinstalling from DESTDIR, determining if a port is already installed in
    DESTDIR, etc. are implemented by chrooting those specific commands.

  * Text changes to reflect that we are installing to DESTDIR, removing from
    DESTDIR, etc.

Note, that this change implements only the infrastructure support.  Ports should
also be implemented to respect DESTDIR.

20060717:
AUTHOR: portmgr@FreeBSD.org
The following functional changes were made to bsd.port.mk:

  * Introduces IA32_BINARY_PORT for certain cases where a given port fetches
    and installs compiled i386 binaries.

  * Adds some infrastructure for support of 32-bit i386 apps on amd64.

  * Again fix USE_LDCONFIG logic to make it work with non-default PREFIX to
    fix bug introduced in previous revision.

20060705:
AUTHOR: portmgr@FreeBSD.org
The following functional changes were made to bsd.port.mk:

  * Change all bogus uses of BROKEN to IGNORE.  (Note: the BROKEN_WITH_*
    forms are retained for compatibility but deprecated.)  Internally,
    these variables all had already set IGNORE, not BROKEN, so they
    were really misnomers.

  * Fix USE_LDCONFIG with non-default PREFIX.

  * Fix DESKTOP_ENTRIES processing on 4.x.

  * Add 'make missing' to show missing dependencies.

  * Add bsd.scons.mk and bsd.wx.mk.

20060616:
AUTHOR: portmgr@FreeBSD.org
  The patch in ports/95841 to get rid of remnants of FORCE_PKG_REGISTER
  was backed out because it broke kdelibs3*.

20060616:
AUTHOR: portmgr@FreeBSD.org

textproc/cdiff/src/cdiff.sh
#!/bin/sh
if [ "$1" = "-m" ]; then
	shift;
	CODES="4:7:1:1"
else
	if [ "$CDIFFCOLORS" != "" ]; then
		CODES=$CDIFFCOLORS
	else
		CODES="1:34:31:35"
	fi
fi
(if [ "$1" != "" ]; then
    while [ "$1" != "" ]; do
    	case $1 in
    	http:*|https:*|ftp:*|file:*)
            fetch -q -o - $1
	    ;;
    	*)
            zcat -f $1
	    ;;
        esac
	shift;
    done
else
	cat
fi) | perl -p -mconstant=CODES,${CODES} -e '
($at, $m, $p, $c) = split /:/, CODES;
while (s/\t/" " x (8 - ((length($`)-1) % 8))/e) {}
s/^(\+.*)$/\033[${p}m$1\033[0m/o;
s/^(-.*)$/\033[${m}m$1\033[0m/o
	if !s/^(--- \d+,\d+ ----.*)$/\033[1m$1\033[0m/o;
s/^(\*\*\* \d+,\d+ *\*\*\*.*)$/\033[1m$1\033[0m/o;
s/^(\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*)$/\033[1m$1\033[0m/o;
s/^(!.*)$/\033[${c}m$1\033[0m/o;
s/^(@@.*$)/\033[${at}m$1\033[0m/o;
if (/^1.(\d+)(\s+\(\w+\s+\d{2}-\w{3}-\d{2}\):\s)(.*)/) {
	$lastcol = $lastcol || 0;
	$lastcol++ if defined($lastrev) && $lastrev != $1;
	$lastrev = $1;
	$lastcol %= 6;
	$_ = "\033[3" . ($lastcol + 1) . "m1.$1$2\033[0m$3\n";
}
' | less -R
